include/astshim/FrameDict.h

```
@@ -0,0 +1,310 @@
+
+    @throws std::invalid_argument if `frame` has a non-empty domain and this FrameDict already
+    contains a Frame with that domain
+    */
+    void addFrame(std::string const &domain, Mapping const &map, Frame const &frame) {
+        addFrame(getIndex(domain), map, frame);
+    }
+
+    /**
+    Get a list of the domain names for all contained Frames that have non-empty domain names.
+
+    The order of the names is unspecified.
+    */
+    std::vector<std::string> getAllDomains() const {
+        std::vector<std::string> domains;
+        domains.reserve(_domainIndexDict.size());
+        for (auto const &item : _domainIndexDict) {
+            domains.emplace_back(item.first);
+        }
+        return domains;
+    }
+
+    using FrameSet::getFrame;
+
+    /**
+    Get a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    std::shared_ptr<Frame> getFrame(std::string const &domain, bool copy = true) const {
+        return FrameSet::getFrame(getIndex(domain), copy);
+    }
+
+    using FrameSet::getMapping;
+
+    //@{
+    /**
+    Get a mapping from one frame to another, with one or both frames specified by domain
+
+    @throw std::range_error if no frame found with the specified from or to domain
+    */
+    std::shared_ptr<Mapping> getMapping(int from, std::string const &to) const {
+        return FrameSet::getMapping(from, getIndex(to));
+    }
+    std::shared_ptr<Mapping> getMapping(std::string const &from, int to) const {
+        return FrameSet::getMapping(getIndex(from), to);
+    }
+    std::shared_ptr<Mapping> getMapping(std::string const &from, std::string const &to) const {
+        return FrameSet::getMapping(getIndex(from), getIndex(to));
+    }
+    //@}
+
+    /**
+    Get the index of a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    int getIndex(std::string const &domain) const {
```

Again, `range_error` is inappropriate here. I guess it's debatable whether `getIndex` should throw `out_of_range` like `getFrame` and `getMapping` should, or whether it should throw `invalid_argument` when called in isolation.

---

I'm really glad you caught this because it raises a better error in Python